Content.Server/Kudzu/SpreaderSystem.cs

```diff
@@ -0,0 +1,131 @@
+using System.Collections.Generic;
+using System.Linq;
+using Content.Server.Atmos.Components;
+using Content.Shared.Atmos;
+using Content.Shared.Movement.EntitySystems;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Map;
+using Robust.Shared.Maths;
+using Robust.Shared.Random;
+
+namespace Content.Server.Kudzu;
+
+// Future work includes making the growths per interval thing not global, but instead per "group"
+public class SpreaderSystem : EntitySystem
+{
+    /// <summary>
+    /// Maximum number of edges that can grow out every interval.
+    /// </summary>
+    private const int GrowthsPerInterval = 1;
+
+    private float _accumulatedFrameTime = 0.0f;
+
+    private readonly HashSet<EntityUid> _edgeGrowths = new ();
+
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<SpreaderComponent, ComponentAdd>(SpreaderAddHandler);
+    }
+
+    private void SpreaderAddHandler(EntityUid uid, SpreaderComponent component, ComponentAdd args)
+    {
+        if (component.Enabled)
+            _edgeGrowths.Add(uid); // ez
+    }
+
+    public void UpdateNearbySpreaders(EntityUid blocker, AirtightComponent comp)
+    {
+        if (!EntityManager.TryGetComponent<TransformComponent>(blocker, out var transform))
+            return; // how did we get here?
+
+        if (!_mapManager.TryGetGrid(transform.GridID, out var grid)) return;
+
+        for (var i = 0; i < Atmospherics.Directions; i++)
+        {
+            var direction = (AtmosDirection) (1 << i);
+            if (!comp.AirBlockedDirection.IsFlagSet(direction)) continue;
+
+            foreach (var ent in grid.GetInDir(transform.Coordinates, direction.ToDirection()))
+            {
+                if (EntityManager.TryGetComponent<SpreaderComponent>(ent, out var s) && s.Enabled)
+                    _edgeGrowths.Add(ent);
+            }
+        }
+    }
+
+    public override void Update(float frameTime)
+    {
+        _accumulatedFrameTime += frameTime;
+
+        if (!(_accumulatedFrameTime >= 1.0f))
+            return;
+
+        _accumulatedFrameTime -= 1.0f;
+
+        var growthList = _edgeGrowths.ToList();
+        _robustRandom.Shuffle(growthList);
```

Comment:
Might be a problem on larger stations but I guess we'll see (doing a shuffle was the most expensive part of the old physics engine).

Reply:
This would be solved if the engine had a IRobustRandom.ShuffleStream(IList<T>), which could be implemented with a range 0..N periodic pseudo-random function of period N (no repeats).
Such things exist (Xorshift32+ is that, with range 0..uint.Max and period uint.Max, no repeats), so would just have to find a fitting one.